vncviewer/touch.cxx

```diff
@@ -0,0 +1,324 @@
+    // simplify all the code handling the masks
+    if (num_masks > 1) {
+      vlog.error(_("Window 0x%08lu has more than one X Input 2 event mask"), iter->first);
+      continue;
+    }
+
+    newmask.deviceid = curmasks[0].deviceid;
+
+    assert(newmask.mask_len >= curmasks[0].mask_len);
+    memcpy(newmask.mask, curmasks[0].mask, curmasks[0].mask_len);
+    if (enable)
+      XISetMask(newmask.mask, XI_TouchOwnership);
+    else
+      XIClearMask(newmask.mask, XI_TouchOwnership);
+
+    XISelectEvents(fl_display, iter->first, &newmask, 1);
+
+    XFree(curmasks);
+  }
+}
+
+int x11_grab_pointer(Window window)
+{
+  int status;
+
+  if (handlers.count(window) == 0) {
+    vlog.error(_("Invalid window 0x%08lu specified for pointer grab"), window);
+    return BadWindow;
+  }
+
+  // We need to remove XI_TouchOwnership from our event masks while
+  // grabbing as otherwise we will get double events (one for the grab,
+  // and one because of XI_TouchOwnership) with no way of telling them
+  // apart. See XInputTouchHandler constructor for why we use this
+  // event.
+  x11_change_touch_ownership(false);
+
+  status = dynamic_cast<XInputTouchHandler*>(handlers[window])->grabPointer();
+
+  if (status != XIGrabSuccess)
+    x11_change_touch_ownership(true);
+
+  return status;
+}
+
+void x11_ungrab_pointer(Window window)
+{
+  if (handlers.count(window) == 0) {
+    vlog.error(_("Invalid window 0x%08lu specified for pointer grab"), window);
+    return;
+  }
+
+  dynamic_cast<XInputTouchHandler*>(handlers[window])->ungrabPointer();
+
+  // Restore XI_TouchOwnership now that the grab is gone
+  x11_change_touch_ownership(true);
+}
+#endif
+
+#if !defined(WIN32) && !defined(__APPLE__)
+extern void dbgEvent(const XEvent& event);
+#endif
+
+static int handleTouchEvent(void *event, void *data)
+{
+#if defined(WIN32)
+  MSG *msg = (MSG*)event;
+
+  // Don't proceed if this isn't a touch device
+  if (!GetSystemMetrics(SM_DIGITIZER)) {
+    return 0;
+  }
+
+  // If window is registered as touch we can not receive gestures
+  if (IsTouchWindow(msg->hwnd, NULL)) {
+    vlog.error(_("Window is registered for touch instead of gestures"));
+    return 0;
+  }
+
+  // Trigger on the first WM_PAINT event. We can't use WM_CREATE events
+  // since FLTK's system handlers trigger before WndProc.
+  // WM_CREATE events are sent directly to WndProc.
+  if (msg->message == WM_PAINT && handlers.count(msg->hwnd) == 0) {
+    handlers[msg->hwnd] = new win32TouchHandler(msg->hwnd);
+    // We need a special hook-in to be able to handle gestures.
+    // Again, we can't rely on FLTK's system handler won't work here.
+    // WM_GESTURE events are generated by WndProc for unhandled WM_POINTER
+    // events. FLTK's system handlers trigger before WndProc.
+    if (!SetWindowSubclass(msg->hwnd, &win32_gestureProc, 1, 0)) {
+      vlog.error(_("Failed to insert gesture handler on window"));
+      return 1;
+    }
```

Comment:
Hide this in the constructor instead?

Reply:
I assume you meant in the Win32 class constructor? As discussed I had to keep this subclass proc function in touch.cxx.